src/eduid_userdb/tests/test_dashboard_user.py

```diff
@@ -27,3 +31,35 @@ def test_verify_mail_and_set_as_primary(self):
         # Restore the old mail settings for other tests
         user.set_mail_aliases(old_mail_aliases)
         user.set_mail(old_mail)
+
+
+class TestPdataUser(TestCase):
+    def test_proper_user(self):
+        userdata = copy.deepcopy(NEW_USER_EXAMPLE)
+        user = DashboardUser(data=userdata)
+        self.assertEqual(user.user_id, userdata['_id'])
+        self.assertEqual(user.eppn, userdata['eduPersonPrincipalName'])
+
+    def test_proper_new_user(self):
+        userdata = copy.deepcopy(NEW_USER_EXAMPLE)
+        userid = userdata.pop('_id')
+        eppn = userdata.pop('eduPersonPrincipalName')
+        passwords = CredentialList(userdata['passwords'])
+        user = DashboardUser.construct_user(_id=userid, eppn=eppn, passwords=passwords)
+        self.assertEqual(user.user_id, userid)
+        self.assertEqual(user.eppn, eppn)
+
+    def test_missing_id(self):
+        userdata = copy.deepcopy(NEW_USER_EXAMPLE)
+        userid = userdata.pop('_id')
+        eppn = userdata.pop('eduPersonPrincipalName')
+        passwords = CredentialList(userdata['passwords'])
+        user = DashboardUser.construct_user(eppn=eppn, passwords=passwords)
+        self.assertNotEqual(user.user_id, userid)
+
+    def test_missing_eppn(self):
+        userdata = copy.deepcopy(NEW_USER_EXAMPLE)
+        userid = userdata.pop('_id')
+        userdata.pop('eduPersonPrincipalName')
+        with self.assertRaises(TypeError):
```

Comment:
A missing mandatory value sounds more like a ValueError or a UserMissingData than a TypeError to me.

Reply:
I'm raising now a UserMissingData here now, but I'm not so sure, TypeError is what Python raises when a method misses a mandatory argument, and `construct_user` require(d) an `eppn` arg.

I think that the problem here is the test(s), it hasn't changed since the point at which I was aiming for `User.construct_user(**user.to_dict())`, and at first sight it seemed that it still tested the right thing, but it doesn't exactly.

I have already pushed a commit raising UserMissingData, but perhaps the correct thing to do here is to fix the tests.

What do you think?